tests/acceptance/serverd-multi-versions.sh

```diff
@@ -0,0 +1,122 @@
+#!/bin/bash
+
+PWD=`pwd`
+logdir=$PWD/serverd-multi-versions-logs
+statsdir=$PWD/serverd-multi-versions-stats
+
+mkdir -p hub
+mkdir -p client
+mkdir -p $logdir
+mkdir -p $statsdir
+cat >server.sh <<EOF
+#!/bin/sh
+
+LD_LIBRARY_PATH=$PWD/hub/var/cfengine/lib CFENGINE_TEST_OVERRIDE_EXTENSION_LIBRARY_DIR=$PWD/hub/var/cfengine/lib CFENGINE_TEST_OVERRIDE_WORKDIR=$PWD/work_here $PWD/hub/var/cfengine/bin/cf-serverd "\$@"
+EOF
+cat >agent.sh <<EOF
+#!/bin/sh
+LD_LIBRARY_PATH=$PWD/agent/var/cfengine/lib CFENGINE_TEST_OVERRIDE_EXTENSION_LIBRARY_DIR=$PWD/agent/var/cfengine/lib CFENGINE_TEST_OVERRIDE_WORKDIR=$PWD/work_here $PWD/agent/var/cfengine/bin/cf-agent "\$@"
+EOF
+chmod a+x *.sh
+sed -i "s%\$(sys.cf_agent)%$PWD/agent.sh%g;s%\$(sys.cf_serverd)%$PWD/server.sh%g" tests/acceptance/run_with_server.cf.sub
+chmod -R 700 tests
+
+mkdir -p work_stub/inputs/
+cp -a $PWD/tests/acceptance work_stub/inputs/
+rm -rf work_stub/inputs/acceptance/0*
+mkdir -p work_stub/bin/
+touch work_stub/bin/cf-promises
+chmod a+x work_stub/bin/cf-promises
+
+sudo hostname localhost
+# Ensure reverse hostname resolution is correct and 127.0.0.1 is always 'localhost'.
+# There's no nice shell command to test it but this one:
+# python -c 'import socket;print socket.gethostbyaddr("127.0.0.1")'
+sudo sed -i -e '1s/^/127.0.0.1 localhost localhost.localdomian\n/' /etc/hosts
+
+# downloads $3 (if provided), saves it as $2 and extracts it to $1
+# if $2 not provided - installs currently built version
+function install(){
+    test -z "$1" && return 1
+    rm -rf $PWD/$1
+    if test -z "$2"
+    then
+        make --silent DESTDIR=$PWD/$1 install
+        mkdir -p $PWD/$1/var
+        mv $PWD/$1/$HOME/cf_install $PWD/$1/var/cfengine
+    else
+        test -z "$3" -o -f $2 || wget $3 -O $2
+        dpkg -x $2 $PWD/$1
+    fi
+}
+
+function testit(){
+    echo "===== $1"
+    $PWD/server.sh -V
+    $PWD/agent.sh -V
+    ls tests/acceptance/16_cf-serverd/serial/*.cf | grep -v -f tests/acceptance/serverd-multi-versions-blacklist.txt | sed 's%^tests/%%' | while read file
+    do
+        logfilename="$1-$(basename $file)"
+        logfile="$logdir/$logfilename"
```

Comment:
I **love** that the logs are so easily accessible! Looking forward to seeing this in the classic acceptance tests.

Reply:
On another note, the logs include only the client, and include messages like the following:

> R: /home/travis/build/cfengine/core/server.sh was run in debug mode, the logs will be in /home/travis/build/cfengine/core/work_here/server-debug.localhost_open.srv.log

Can we include the serverd logs in the archive?

Reply:
And on another note, the test passes, but I see complaints of "abnormal termination":

> Q: "...e/agent.sh -KIf": R: /home/travis/build/cfengine/core/work_here/inputs/acceptance/16_cf-serverd/serial/001.cf Pass
>    error: Finished script '/home/travis/build/cfengine/core/server.sh -Kdf /home/travis/build/cfengine/core/work_here/inputs/acceptance/16_cf-serverd/serial/localhost_open.srv > /home/travis/build/cfengine/core/work_here/server-debug.localhost_open.srv.log 2>&1' - failed (abnormal termination)

Any ideas why?

Reply:
serverd logs are included in *.workdir dir in archive for failed tests. I thought that it's a waste to include them for passed tests.

Regarding `error: Finished script '...' - failed (abnormal termination)` - it looks like expected CFEngine output. When we terminate cf-serverd, we send two signals to same processes: SIGTERM and SIGKILL [(run_with_server.cf.sub#L106)](https://github.com/cfengine/core/blob/master/tests/acceptance/run_with_server.cf.sub#L106). If server shuts down cleanly after first signal - the second doesn't reach its target and prints error message like "process PID not found"; if server does not shut down nicely after first signal - it gets killed and above error message gets printed. That's my understanding, at least.

Reply:
> Regarding error: Finished script '...' - failed (abnormal termination) - it looks like expected CFEngine output. When we terminate cf-serverd, we send two signals to same processes: SIGTERM and SIGKILL (run_with_server.cf.sub#L106). If server shuts down cleanly after first signal - the second doesn't reach its target and prints error message like "process PID not found"; if server does not shut down nicely after first signal - it gets killed and above error message gets printed. That's my understanding, at least.

Ugly, but you are right. I have always wanted to insert a delay between the 2 signals: https://tracker.mender.io/browse/CFE-2207